[AMD] Hoist Q out of the loop for FA optimization

oplavsic opened the pull request:
Move writing to LDS and reading from LDS right after the loading of a tensor from global memory. This PR does reordering by considering 2 possible patterns depending on whether writing to LDS is done using an optional local_alloc argument or a local_store instruction: 1) load -> local_alloc -> local_store -> local_load, 2) load -> local_alloc -> local_load.

sjw36 commented on third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp:
```diff
@@ -97,6 +98,51 @@ class TritonAMDGPUReorderInstructionsPass
+          Operation *loadOp = localStore->getOperand(0).getDefiningOp();
+
+          if (isa<tt::LoadOp>(loadOp)) {
+            localAlloc->moveAfter(loadOp);
```
What about all the pointer logic that these ops depend on? 
You must capture/move the cone of logic for all operands of the localStore. 
The `local_store` may also be in the loop and it may not be valid to hoist it.

sjw36 replied:
Check localAlloc->getBlock() == localStore->getBlock() == loadOp->getBlock() to verify in the same block.

And check below where the logic cone is captured (line 130).

antiagainst replied:
+1

sjw36 reviewed (commented):
This change seems to undo some of the ordering that the new pipeliner sets. For instance, the pipeliner will issue the global loads for all buffers first, then write those to shared memory. Is there a perf benefit to interleaving these (ie. load(A)->local_store(A) - load(B)->local_store(B)?

zhanglx13 commented:
I think this is for the Q tensor in FA kernels. And for Q the load is outside of the loop so stream-pipeline will not touch it. 
Then if you see local_alloc/load/store inside the loop, you can find its definingOp. And if the definingOp is outside the loop, instead of inside the loop or being a blockArg, we know this is for Q. 
More generally, such local_xxx can be hoisted out of the loop since we don't want to 1) use more LDS, and 2) load from LDS at every iteration of the loop.

sjw36 commented:
> I think this is for the Q tensor in FA kernels. And for Q the load is outside of the loop so stream-pipeline will not touch it. Then if you see local_alloc/load/store inside the loop, you can find its definingOp. And if the definingOp is outside the loop, instead of inside the loop or being a blockArg, we know this is for Q. More generally, such local_xxx can be hoisted out of the loop since we don't want to 1) use more LDS, and 2) load from LDS at every iteration of the loop.

Ok, all the logic must be checked to verify loop invariance. This can be checked when gathering the cone.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp:
```diff
@@ -97,6 +98,51 @@ class TritonAMDGPUReorderInstructionsPass
+    //    %3 = local_store %1, %2
+    //    %4 = local_load %2
+    m.walk([&](ttg::LocalLoadOp op) {
+      Operation *localLoad = (Operation *)op;
```
Do we need this definition? You can just name the above `op` as `localLoad` and it [supports `->`](https://github.com/llvm/llvm-project/blob/5b95c9e/mlir/include/mlir/IR/OpDefinition.h#L105)?

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp:
```diff
@@ -97,6 +98,51 @@ class TritonAMDGPUReorderInstructionsPass
+      Operation *localLoad = (Operation *)op;
+      Operation *localAlloc = localLoad->getOperand(0).getDefiningOp();
+
+      if (!localAlloc || !isa<ttg::LocalAllocOp>(localAlloc)) {
```
This can be done as

```c++
auto localAlloc = localLoad->getDefiningOp<ttg::LocalAllocOp>();
if (!localAlloc) return;
```

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp:
```diff
@@ -97,6 +98,51 @@ class TritonAMDGPUReorderInstructionsPass
+      }
+
+      // Case when localAlloc has no operands
+      if (localAlloc->getNumOperands() < 1) {
```
Can we check for the # operand == 1 case first so we can do early return for that simpler one?

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp:
```diff
@@ -97,6 +98,51 @@ class TritonAMDGPUReorderInstructionsPass
+
+        // Check if the use is a LocalStoreOp
+        if (isa<ttg::LocalStoreOp>(localStore)) {
+          Operation *loadOp = localStore->getOperand(0).getDefiningOp();
```
Similarly here: `getDefiningOp<tt::LoadOp>()`.

antiagainst reviewed (changes requested):
Thanks! A few nits. Also +1 to what Simon said. I think we are starting to see conflicting needs so we should restructure this pass. I'm fine landing this fix first and then doing that next.

oplavsic commented:
Thanks for detailed review! I addressed @antiagainst code cleanup comments and added some more checks to better verify that the code is affecting only these very specific patterns. @sjw36 I'm not sure I quite follow all of the comments. The way I see it, this code affects only two very particular cases. It should not undo anything that pipeliner has done. If the tensor is pipelined and argument to local_load (or even local_store) is the block argument, this code will bail (on getDefiningOp<SomeOp>(); if(SomeOp)  ... checks that are in the code). So the code should only affect tensors that are not pipelined. Moreover, it affects only cases where localAlloc has 1 or 2 uses (local_load, or local_load and local_store), and again, the code bails if it's not exactly this pattern with this number of uses. So I'm not sure I understand what pointer arithmetic can be broken here. Pipeliner lit tests are passing. Is there some particular test case or config in general you are worried about that I can run and see in more detail (by examining TTGIR) cases that you are worried about?

sjw36 reviewed (commented):
Code reviewed with @oplavsic, the pattern makes sense and captures all necessary ops. 
Minor nits and please add check for specifically moving local_ ops out of loop.

antiagainst commented on third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp:
```diff
@@ -97,6 +98,66 @@ class TritonAMDGPUReorderInstructionsPass
+    //    %4 = local_load %2
+    m.walk([&](ttg::LocalLoadOp localLoad) {
+      auto localAlloc =
+          localLoad->getOperand(0).getDefiningOp<ttg::LocalAllocOp>();
```
I addressed this along a few style nits--need to retrigger the CI after some network issues anyway.
